File: w10-mean/d3-drills-dfs/rb/tree-solution.rb
##
# This class represents a tree
# where each node has an array of children.

class Tree

  ##
  # reads the key of the tree's root
  attr_reader :key

  ## 
  # sets the key of the tree's root node
  # unless intended new key is nil
  # (to remove a node, set the entire node to nil)
  def key=(new_key)
    if !new_key.nil?
      @key = new_key
    end
    @key
  end

  ##
  # sets or reads the array of children for this instance
  attr_accessor :children


  ##
  # Creates a new tree with the root node key specified by 
  # the +key+ param as the @key for the instance
  def initialize(key)
    @key = key
    @children = []
  end

  ##
  # Searches through tree for a single node whose key is equal to
  # the +target_key+ param. Looks through nodes in a depth-first order,
  # by following each path as deep as it can before backtracking.
  # Returns nil if no such node is found.
  def depth_first_search(target_key)
    stack = [{:node => self, :visited => false}]
    while !stack.empty?
      # print out key of each entry in stack to help with debugging
      p stack.map { |entry| "#{entry[:node].key}#{entry[:visited] ? '.' : ''}" }

      # peek at the last entry in the stack
      current_entry = stack[stack.length-1]
      # if we've already visited it once, we're done with it
      if current_entry[:visited]
        stack.pop
      else  # we haven't visited this node yet
        current_node = current_entry[:node]
        current_entry[:visited] = true


        # check if this node has the right key (and return if so)
        if current_node.key == target_key
          return current_node
        end
        # add this node's children to the stack
        current_node.children.each do |child|
          stack.push({:node => child, :visited => false})
        end
      end

    end

    # We finished the while loop (went through all nodes)
    # without finding the node we wanted.
    # implicitly return nil
    nil
  end


  ##
  # Searches through all nodes of the tree, spreading 
  # outward from the root. Looks for any node with key equal
  # to the +target_key+ param.
  def breadth_first_search(target_key)
    queue = [self]
    while !queue.empty?
      # print out key of each node in queue to help with debugging
      p queue.map { |node| node.key }

      # remove the first node from the queue
      current = queue.shift
      # check if this node has the right key (and return if so)
      if current.key == target_key
        return current
      end
      # add this node's children to the queue
      queue = queue + current.children
    end

    # We finished the while loop (went through all nodes)
    # without finding the node we wanted.
    # implicitly return nil
    nil
  end

end

my_tree = Tree.new('Q')
r_node = Tree.new('R')
s_node = Tree.new('S')
t_node = Tree.new('T')
u_node = Tree.new('U')
v_node = Tree.new('V')
w_node = Tree.new('W')

my_tree.children << r_node
my_tree.children << s_node
r_node.children << t_node
r_node.children << u_node
r_node.children << v_node
u_node.children << w_node

#          T 
#        /
#     R  -- U -- W
#   /   \
# Q        V
#   \
#     S 

puts "\n-- expect node U --"
u_result = my_tree.depth_first_search('U')
p u_result
# ["Q"]
# ["Q.", "R", "S"]
# ["Q.", "R", "S."]
# ["Q.", "R"]
# ["Q.", "R.", "T", "U", "V"]
# ["Q.", "R.", "T", "U", "V."]
# ["Q.", "R.", "T", "U"]
# #<Tree:0x007fa95a96f108 @key="U", @children=[#<Tree:0x007fa95a96f018 @key="W", @children=[]>]>

puts "\n-- expect node W --"
w_result = my_tree.depth_first_search('W')
p w_result
# ["Q"]
# ["Q.", "R", "S"]
# ["Q.", "R", "S."]
# ["Q.", "R"]
# ["Q.", "R.", "T", "U", "V"]
# ["Q.", "R.", "T", "U", "V."]
# ["Q.", "R.", "T", "U"]
# ["Q.", "R.", "T", "U.", "W"]
# #<Tree:0x007fa95a96f018 @key="W", @children=[]>


puts "\n-- expect nil -- "
z_result = my_tree.depth_first_search('Z')
p z_result
# ["Q"]
# ["Q.", "R", "S"]
# ["Q.", "R", "S."]
# ["Q.", "R"]
# ["Q.", "R.", "T", "U", "V"]
# ["Q.", "R.", "T", "U", "V."]
# ["Q.", "R.", "T", "U"]
# ["Q.", "R.", "T", "U.", "W"]
# ["Q.", "R.", "T", "U.", "W."]
# ["Q.", "R.", "T", "U."]
# ["Q.", "R.", "T"]
# ["Q.", "R.", "T."]
# ["Q.", "R."]
# ["Q."]
# nil

File: w10-mean/d3-drills-dfs/js/tree-solution.js
// Note: To run only recursive version of depth first search tests from the Terminal 
// use `node tree-solution.js recursive`.
// For iterative only, use `node tree-solution.js iterative`.


// This object type represents a tree
// where each node has an array of children.
var TreeNode = function (key, children){
	this.key = key; 	
	this.children = children || [];	// an array, holds the TreeNodes that we can go to directly from here
};

// Searches through tree for a single node whose key is equal to
// the +target_key+ param. Looks through nodes in a depth-first order,
// by following each path as deep as it can before backtracking.
// Returns null if no such node is found.
TreeNode.prototype.depthFirstSearch	= function(targetKey) {
	var stack = [{node: this, visited: false}];
	var current;

	var addChildToStack = function(child){
		stack.push({node: child, visited: false});
	};

	while (stack.length !== 0) {
		// print out key of each node in stack to help with debugging
		console.log("stack: ", stack.map(function(entry){
			return entry.node.key + (entry.visited ? "." : "");
		}));
		// peek at top item in the stack
		current = stack[stack.length-1];
		if (current.node.key == targetKey){
			return current.node;
		}
		if (current.visited){
			stack.pop();
		} else {
			current.visited = true;
			current.node.children.forEach(addChildToStack);
		}
	}
	return null;
};


TreeNode.prototype.depthFirstSearchRecursive = function(target_key) {
	// current node is this
	console.log("processing ", this.key);
	if (this.key == target_key){
		return this;
	}
	var len = this.children.length;
	var i = 0;
	var childResult = null;

	for (; i<len; i++) {
		// recursively search each child
		childResult = this.children[i].depthFirstSearchRecursive(target_key);
		if (childResult){
			// if child has a node matching the target, can return it
			return childResult;
		}
	}

	// if we reach this point without returning, 
	// neither this node nor any in its subtree had the target key
	console.log("finished ", this.key);
	return null;
};


// Searches through all nodes of the tree, spreading 
// outward from the root. Looks for any node with key equal
// to the targetKey param. Returns null if no such node is found.
TreeNode.prototype.breadthFirstSearch = function(targetKey){
	// set up the queue of nodes to check 
	var queue = [this]; 
	var current;
	while (queue.length !== 0){
		// print out key of each node in queue to help with debugging
		console.log("queue: ", queue.map(function(node){
			return node.key;
		}));

		current = queue.shift(); // remove and return first element of array

		// check if current node key matches the target key
		if (current.key == targetKey){
			return current;
		} else {
			if (current.children){
				queue = queue.concat(current.children);
			}
		}
	}
	// we've gone through the entire tree without finding the key
	return null;
};


// Return all nodes for which the selectionFunction returns true.
TreeNode.prototype.breadthFirstSearchAll = function(selectionFunction){
	// set up the queue of nodes to check 
	var queue = [this]; 
	// set up output array of nodes that match properties
	var output = [];
	var current;
	while (queue.length !== 0){
		// print out key of each node in queue to help with debugging
		console.log("queue: ", queue.map(function(node){
			return node.key;
		}));

		current = queue.shift(); // remove and return first element of array

		// check if selectionFunction returns true for current node 
		if (selectionFunction(current)){
			output.push(current);
		} else {
			if (current.children){
				queue = queue.concat(current.children);
			}
		}
	}
	return output;
};



var myTree = new TreeNode('Q'),
	rNode = new TreeNode('R'),
	sNode = new TreeNode('S'),
	tNode = new TreeNode('T'),
	uNode = new TreeNode('U'),
	vNode = new TreeNode('V'),
	wNode = new TreeNode('W');

myTree.children.push(rNode, sNode);
rNode.children.push(tNode, uNode, vNode);
uNode.children.push(wNode);

//          T 
//        /
//     R  -- U -- W
//   /	  \
// Q        V
//   \
//     S 



//************ ITERATIVE ****************//
if (!process.argv[2] || process.argv[2] == "iterative") {
	console.log("\n-- expect node U --");
	console.log(myTree.depthFirstSearch('U'));  
	// -- expect node U --
	// stack:  [ 'Q' ]
	// stack:  [ 'Q., 'R', 'S' ]
	// stack:  [ 'Q., 'R', 'S.' ]
	// stack:  [ 'Q., 'R' ]
	// stack:  [ 'Q., 'R.', 'T', 'U', 'V' ]
	// stack:  [ 'Q., 'R.', 'T', 'U', 'V.' ]
	// stack:  [ 'Q., 'R.', 'T', 'U' ]
	// { key: 'U', children: [ { key: 'W', children: [] } ] }


	console.log("\n-- expect node W --");
	console.log(myTree.depthFirstSearch('W'));  
	// -- expect node W --
	// stack:  [ 'Q' ]
	// stack:  [ 'Q., 'R', 'S' ]
	// stack:  [ 'Q., 'R', 'S.' ]
	// stack:  [ 'Q., 'R' ]
	// stack:  [ 'Q., 'R.', 'T', 'U', 'V' ]
	// stack:  [ 'Q., 'R.', 'T', 'U', 'V.' ]
	// stack:  [ 'Q., 'R.', 'T', 'U' ]
	// stack:  [ 'Q., 'R.', 'T', 'U.', 'W' ]
	// { key: 'W', children: [] }


	console.log("\n-- expect null --");
	console.log(myTree.depthFirstSearch('Z')); 
	// -- expect null --
	// stack:  [ 'Q' ]
	// stack:  [ 'Q., 'R', 'S' ]
	// stack:  [ 'Q., 'R', 'S.' ]
	// stack:  [ 'Q., 'R' ]
	// stack:  [ 'Q., 'R.', 'T', 'U', 'V' ]
	// stack:  [ 'Q., 'R.', 'T', 'U', 'V.' ]
	// stack:  [ 'Q., 'R.', 'T', 'U' ]
	// stack:  [ 'Q., 'R.', 'T', 'U.', 'W' ]
	// stack:  [ 'Q., 'R.', 'T', 'U.', 'W.' ]
	// stack:  [ 'Q., 'R.', 'T', 'U.' ]
	// stack:  [ 'Q., 'R.', 'T' ]
	// stack:  [ 'Q., 'R.', 'T.' ]
	// stack:  [ 'Q., 'R.' ]
	// stack:  [ 'Q. ]
	// null
}

//*************RECURSIVE********************//
if (!process.argv[2] || process.argv[2] == "recursive") {

	console.log("\n-- expect node U --");
	console.log(myTree.depthFirstSearchRecursive('U'));  
	// processing  Q
	// processing  R
	// processing  T
	// finished  T
	// processing  U
	// { key: 'U', children: [ { key: 'W', children: [] } ] }


	console.log("\n-- expect node W --");
	console.log(myTree.depthFirstSearchRecursive('W'));  
	// processing  Q
	// processing  R
	// processing  T
	// finished  T
	// processing  U
	// processing  W
	// { key: 'W', children: [] }


	console.log("\n-- expect null --");
	console.log(myTree.depthFirstSearchRecursive('Z')); 
	// processing  Q
	// processing  R
	// processing  T
	// finished  T
	// processing  U
	// processing  W
	// finished  W
	// finished  U
	// processing  V
	// finished  V
	// finished  R
	// processing  S
	// finished  S
	// finished  Q
	// null
}
